[TVMC] Allow selecting a subset of tasks to be used in `tvmc tune`

PhilippvK opened the pull request:
This adds a `--tasks` flag to the `tvmc tune` command to filter the lists of tasks to be tuned. See examples below.

## Motivation

- As auto-tuning can be quite time consuming, it is often desirable to cut down the number of tuned tasks in a session.
- If the tuning session was canceled halfway through, it would be a bad idea to start from scratch. Instead continue with the last untuned task
- Some tasks have more impact on the model performance than others, thus we should be able to train some tasks longer than others

## Examples

1. Use `--task list` to show which tasks are available for tuning
```
$ tvmc tune toycar.tflite -o out.txt --task list
Available Tasks for tuning:
  0. Task(func_name=dense_nopack.x86, args=(('TENSOR', (1, 640), 'int16'), ('TENSOR', (128, 640), 'int...
  1. Task(func_name=dense_pack.x86, args=(('TENSOR', (1, 640), 'int16'), ('TENSOR', (128, 640), 'int16...
  2. Task(func_name=dense_nopack.x86, args=(('TENSOR', (1, 128), 'int16'), ('TENSOR', (128, 128), 'int...
  3. Task(func_name=dense_pack.x86, args=(('TENSOR', (1, 128), 'int16'), ('TENSOR', (128, 128), 'int16...
  4. Task(func_name=dense_nopack.x86, args=(('TENSOR', (1, 128), 'int16'), ('TENSOR', (8, 128), 'int16...
  5. Task(func_name=dense_pack.x86, args=(('TENSOR', (1, 128), 'int16'), ('TENSOR', (8, 128), 'int16')...
  6. Task(func_name=dense_nopack.x86, args=(('TENSOR', (1, 8), 'int16'), ('TENSOR', (128, 8), 'int16')...
  7. Task(func_name=dense_pack.x86, args=(('TENSOR', (1, 8), 'int16'), ('TENSOR', (128, 8), 'int16'), ...
  8. Task(func_name=dense_nopack.x86, args=(('TENSOR', (1, 128), 'int16'), ('TENSOR', (640, 128), 'int...
  9. Task(func_name=dense_pack.x86, args=(('TENSOR', (1, 128), 'int16'), ('TENSOR', (640, 128), 'int16...
```

2. Filter the list of tasks to be tuned:

```
# Only tune a single task (index 5)
tvmc tune toycar.tflite -o out.txt --tasks 5

# Tunes tasks starting with index 6
tvmc tune toycar.tflite -o out.txt --tasks "6-"

# Tune tasks 1,4,5,6,8,9
tvmc tune toycar.tflite -o out.txt --tasks "1,4-6,8-"
```

## Tests
I added a basic unit test for the `filter_tasks` utility in `tests/python/driver/tvmc/test_autotuner.py`.

## Open Questions
- ~~While the (truncated) string representations of AutoTVM tasks are quite helpful to pick the correct tasks, using AutoScheduler the tasks can not really be distinguished from each other (only by index). Is there a way to get similar information from AutoScheduler tasks?~~

cc @Mousius @gromero

PhilippvK commented:
CC @leandron @giuseros @Mousius @areusch @tqchen

areusch commented on python/tvm/driver/tvmc/autotuner.py:
```diff
@@ -453,7 +518,37 @@ def tune_model(
+                "\n".join(
+                    [
+                        "  {}. {}".format(
+                            i, task if len(str(task)) < 100 else str(task)[:97] + "..."
```
in this case does this make the task useless?

PhilippvK replied:
Sorry, but I do not get the question… This truncates the string representation of each task to 100 characters which is typically enough for AutoTVM task to show the operator name and input shape.

Or do you mean what would happen in a situation where `list` as well as a filter expression is passed by the user?

areusch replied:
oh sorry, i meant to clarify this point before sending and then i forgot to delete it after realizing it was moot. originally i read the diff before the full description and thought that this was more like a filter-by-name command-line option. so, please ignore :)

areusch commented on python/tvm/driver/tvmc/autotuner.py:
```diff
@@ -293,9 +298,66 @@ def drive_tune(args):
+                lhs, rhs = item.split("-")[:2]
+                if len(lhs) == 0:
+                    assert len(rhs) > 0
+                    assert isinstance(rhs, str)
```
not sure we need these asserts, since item should already be a str right?

areusch commented on python/tvm/driver/tvmc/autotuner.py:
```diff
@@ -293,9 +298,66 @@ def drive_tune(args):
+            do_filter = True
+            if "-" in item:
+                lhs, rhs = item.split("-")[:2]
+                if len(lhs) == 0:
```
i think you could maybe make this a bit more concise:
lhs = 
```suggestion
                lhs = int(lhs) if lhs else 0
                rhs = int(rhs) if rhs else len(tasks) - 1
                assert 0 <= lhs < len(tasks), "explanation"
                assert 0 <= rhs < len(tasks), "explanation"
```

areusch commented on tests/python/driver/tvmc/test_autotuner.py:
```diff
@@ -182,3 +183,23 @@ def test_tune_rpc_tracker_parsing(mock_load_model, mock_tune_model, mock_auto_sc
+    filter_tasks(list(range(10)), "10")
+    filter_tasks(list(range(10)), "5,10")
+    filter_tasks(list(range(10)), "1-10")
+    filter_tasks(list(range(10)), "-10")
```
do you mind adding the customary

```
if __name__ == "__main__":
  tvm.testing.main()
```
here?

areusch reviewed (commented):
thanks @PhilippvK, this looks like a great addition! left a couple comments here and there

areusch commented:
@PhilippvK i think this one needs a rebase

PhilippvK commented:
Short update:
- With my latest changes I can now print useful task names for AutoScheduler as well (using `task.desc` instead of `str(task)`)
- I added a mention of the `--task` argument in the TVMC CLI tutorial

areusch reviewed (approved):
thanks @PhilippvK, could you address the lint issue?

PhilippvK commented:
@areusch 

As I have to retrigger the CI anyway so I want to propose another small addition:

For AutoTVM, in addition to the task description, I could also print `len(task.config_space)` to give an idea on how large the tuning search space is for a specific task, which can help a lot to estimate the required number of tuning trials.

Do you think this would be a great addition to `tvmc tune --task list`?

PhilippvK commented:
@areusch Can I have an update on this?

leandron commented:
@tvm-bot rerun

areusch commented:
sorry @PhilippvK I missed this one. feel free to ping me on discord if that happens again. I retriggered the CI, I'm good for @leandron to merge this if it passes.

> Example:
> 
> Available Tasks for tuning:
>   0. Task(func_name=dense_nopack.x86, args=(('TENSOR', (1, 640), 'int16'), ('TENSOR', (128, 640)... (len=640)
>   1. Task(func_name=dense_pack.x86, args=(('TENSOR', (1, 640), 'int16'), ('TENSOR', (128, 640), ... (len=35)
> Do you think this would be a good addition to tvmc tune --task list?

Sure that seems like a good addition!

PhilippvK commented:
Rebased and CI should be fixed now @leandron @areusch.

leandron commented:
Sorry @PhilippvK. It took me a while to catch up with this. Can you rebase and then we move to get this merged?

PhilippvK commented:
@leandron Rebased on `upstream/main`

areusch commented:
@tvm-bot rerun

PhilippvK commented:
Rebased ontop of `main` and CI errors seem to be fixed. Can we merge this @areusch @leandron ?

PhilippvK commented:
Any further comments @areusch @leandron?

PhilippvK commented:
Rebased to fix conflict + I added 2 small AutoTVM related improvements:
1. If #trials is less then, #tasks, tune each task at least once, instead of failing all of them.
2. If the config space of a single task is smaller than the determined number of trials per task, use the smaller values as bound in the progress bar.

lhutton1 commented on tests/python/driver/tvmc/test_autotuner.py:
```diff
@@ -207,3 +209,27 @@ def test_autotune_pass_context(mock_pc, onnx_mnist, tmpdir_factory):
+    filter_tasks(list(range(10)), "0,4-5,9,list") == ([0, 4, 5, 9], True)
+
+
+@pytest.mark.xfail
```
XFail is normally used to mark a test that is "unexpectedly failing", which doesn't seem like the case here. If the test suddenly starts passing it won't be flagged in CI. I think it would be better to rewrite this test using `pytest.raises` e.g. something like:
```
@pytest.mark.parameterize("value,err_msg", [("10", "my expected error message"), ("5,10", "my expected error message 2"), ...])
def test_filter_tasks_invalid(value, err_msg):
    with pytest.raises(AssertionError, match=err_msg):
        filter_tasks(list(range(10)), value)
    
```

lhutton1 commented on tests/python/driver/tvmc/test_autotuner.py:
```diff
@@ -207,3 +209,27 @@ def test_autotune_pass_context(mock_pc, onnx_mnist, tmpdir_factory):
     assert mock_pc.call_args_list[0][1]["opt_level"] == 3
+
+
+def test_filter_tasks_valid():
```
Nit: it would be helpful to parameterize these tests similar to below, so failures are reported separately in the CI log

lhutton1 commented on python/tvm/driver/tvmc/autotuner.py:
```diff
@@ -290,10 +295,82 @@ def drive_tune(args):
+        return text if len(text) < length else text[: length - 3] + "..."
+
+    print(
+        "\n".join(
```
nit: is it possible to write a test to check this output?

lhutton1 reviewed (approved):
I just noticed this PR and think it looks very helpful, thanks! I had a couple of nits around the testing, but given this PR has been rebased a few times I'm happy for them to be fixed in a follow-up, let me know :)

lhutton1 commented:
@tvm-bot rerun

PhilippvK commented:
@lhutton1 Hey, I implemented the proposed changes and the CI passed. Feel free to merge after quickly having a look at the refactored unit tests.

lhutton1 reviewed (approved):
Thanks @PhilippvK, LGTM!

lhutton1 commented:
Thanks @PhilippvK @areusch @leandron!
